src/metabase/driver/sql/parameters/substitution.clj

```diff
@@ -262,30 +262,42 @@
   (assert (:id field) (format "Why doesn't Field have an ID?\n%s" (u/pprint-to-str field)))
   (letfn [(prepend-field [x]
             (update x :replacement-snippet
-                    (partial str (field->identifier driver field param-type) " ")))]
+                    (partial str (field->identifier driver field param-type) " ")))
+          (->honeysql [form]
+            (sql.qp/with-driver-honey-sql-version driver
+              (sql.qp/->honeysql driver form)))]
     (cond
       (params.ops/operator? param-type)
       (let [[snippet & args]
-            (sql.qp/with-driver-honey-sql-version driver
-              (as-> (assoc params :target [:template-tag (field->clause driver field param-type)]) form
-                (params.ops/to-clause form)
-                (mbql.u/desugar-filter-clause form)
-                (qp.wrap-value-literals/wrap-value-literals-in-mbql form)
-                (sql.qp/->honeysql driver form)
-                (sql.qp/format-honeysql driver form)))]
+            (->> (assoc params :target [:template-tag (field->clause driver field param-type)])
+                 params.ops/to-clause
+                 mbql.u/desugar-filter-clause
+                 qp.wrap-value-literals/wrap-value-literals-in-mbql
+                 ->honeysql
+                 (honeysql->replacement-snippet-info driver))]
         {:replacement-snippet snippet, :prepared-statement-args (vec args)})
-      ;; convert date ranges to DateRange record types
-      (params.dates/date-range-type? param-type) (prepend-field
-                                                  (date-range-field-filter->replacement-snippet-info driver value))
+
+      (and (params.dates/date-type? param-type)
+           (re-matches params.dates/date-exclude-regex value))
+      (let [field-clause (field->clause driver field param-type)]
+        (->> (params.dates/date-string->filter value field-clause)
+             mbql.u/desugar-filter-clause
+             qp.wrap-value-literals/wrap-value-literals-in-mbql
+             ->honeysql
+             (honeysql->replacement-snippet-info driver)))
```

Comment:
This is the block of code that fixes the bug. "Exclude..." type params never worked with native query template tags because the next condition on [line 281](https://github.com/metabase/metabase/pull/30760/files#diff-b545fa8a68713ab8d1171bc5fb528b1ff68042490d9a25f28e480ba4c04dc3c5L281) would evaluate to `true`, and the value (e.g. "exclude-months-Jan") would be passed to `date-range-field-filter->replacement-snippet-info`, which would blow up because it's not a date range. So the solution is to match against exclude type parameters and convert them to an MBQL filter clause with `params.dates/date-string->filter`.